src/ipa_pc/mod.rs

```diff
@@ -1151,30 +781,136 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
+                ).to_owned()))?
+            }
+
+            let polynomial = labeled_polynomial.polynomial();
+            let is_hiding = labeled_polynomial.is_hiding();
+            let randomness = labeled_randomness.randomness();
+
+            let p_len = polynomial.coeffs.len();
+            let segments_count = std::cmp::max(1, p_len / key_len + if p_len % key_len != 0 { 1 } else { 0 });
+
+            if is_hiding {
+                has_hiding = true;
+            }
+
+            let mut polynomial_lc = Polynomial::zero();
+            let mut rand_lc = G::ScalarField::zero();
+
+            // Compute the query-point dependent linear combination of the segments,
+            // both for witnesses, commitments (and their randomnesses, if hiding)
+            for i in 0..segments_count {
+                let is = i * key_len;
+                let poly_single = Polynomial::from_coefficients_slice(
+                    &polynomial.coeffs[i * key_len..core::cmp::min((i + 1) * key_len, p_len)]
+                );
+                // add x^{i*|S|}* p_i(X) of the segment polynomial p_i(X)
+                polynomial_lc += (point.pow(&[is as u64]), &poly_single);
+                if has_hiding {
+                    let rand_single = randomness.rand[i];
+                    rand_lc += &(point.pow(&[is as u64]) * &rand_single);
+                }
+            }
+
+            // add segment linear combination to overall combination,
+            // both for witnesses, commitments (and their randomnesses, if hiding)
+            combined_polynomial += (cur_challenge, &polynomial_lc);
+            if has_hiding {
+                combined_rand += &(cur_challenge * &rand_lc);
+            }
 
-            n /= 2;
+            // next power of lambda
+            cur_challenge = cur_challenge * &lambda;
         }
 
-        end_timer!(proof_time);
+        end_timer!(combine_time);
 
-        Ok(Proof {
-            l_vec,
-            r_vec,
-            final_comm_key: comm_key[0],
-            c: coeffs[0],
-            hiding_comm: hiding_commitment,
-            rand: combined_rand,
-        })
+        Self::open(
+            ck,
+            combined_polynomial,
+            point,
+            has_hiding,
+            combined_rand,
+            fs_rng,
+            rng,
+        )
     }
 
-    /// The multi point multi poly opening proof from [[BDFG2020]](https://eprint.iacr.org/2020/081) 
+    /// The multi point multi poly opening proof from [[BDFG2020]](https://eprint.iacr.org/2020/081)
     /// CAUTION: This is a low-level function which assumes that the statement of the
-    /// opening proof (i.e. commitments, query point, and evaluations) is already bound 
+    /// opening proof (i.e. commitments, query point, and evaluations) is already bound
     /// to the internal state of the Fiat-Shamir rng.
-    fn batch_open_individual_opening_challenges<'a>(
+    fn multi_point_multi_poly_open<'a>(
```

Comment:
same here: let us move it to a default implementation.

Reply:
Done